test/test_sparse.py

```diff
@@ -2678,6 +2679,103 @@ def sparse_log(x):
+            assert r[-1] == nnz
+            return r
+
+        def sparse_mm(a, b):
+            if a.shape[1] != b.shape[0]:
+                print("Incompatible matrices")
+                return None
+            n, p = a.shape
+            p, m = b.shape
+            indices_a = a._indices()
+            values_a = a._values()
+            indices_b = b._indices()
+            values_b = b._values()
+            nnz1 = len(indices_a[0])
+            nnz2 = len(indices_b[0])
+
+            if a.is_coalesced() and b.is_coalesced():
+                r2 = _indices2csr(indices_b[0], b.shape[0])
+                d = defaultdict(values_b.numpy().dtype.type)
+                for n1 in range(nnz1):
+                    for n2 in range(r2[indices_a[1][n1]], r2[indices_a[1][n1] + 1]):
+                        d[indices_a[0][n1].item(), indices_b[1][n2].item()] += values_a[n1] * values_b[n2]
+
+            else:
+                d = defaultdict(values_b.numpy().dtype.type)
+                for n1 in range(nnz1):
+                    for n2 in range(nnz2):
+                        if indices_b[0][n2] == indices_a[1][n1]:
+                            d[indices_a[0][n1].item(), indices_b[1][n2].item()] += values_a[n1] * values_b[n2]
+
+            i3 = []
+            j3 = []
+            values = []
+            for i, j in sorted(d):
+                i3.append(i)
+                j3.append(j)
+                values.append(d[i, j])
+            return torch.sparse_coo_tensor(torch.tensor([i3, j3]), torch.tensor(values), (n, m))
+
+        def test_op(sparse_dims, nnz, shape_a, shape_b):
+            a, i_a, v_a = self._gen_sparse(sparse_dims, nnz, shape_a)
+            b, i_b, v_b = self._gen_sparse(sparse_dims, nnz, shape_b)
+            a_dense = a.to_dense()
+            b_dense = b.to_dense()
+            r1 = a_dense.matmul(b_dense)
+
+            # python implementation
+            r2 = sparse_mm(a, b)
+            self.assertEqual(r1, r2.to_dense())
+            self.assertEqual((r2._values() != 0).all(), True)
+
+            # cpp implementation
+            r2 = torch.sparse.mm(a, b)
+            self.assertEqual(r1, r2.to_dense())
+
+            # check autograd support on sparse matmul
+            a, i_a, v_a = self._gen_sparse(sparse_dims, nnz, shape_a)
+            b, i_b, v_b = self._gen_sparse(sparse_dims, nnz, shape_b)
+            a.requires_grad = True
+            b.requires_grad = True
+            a_dense = a.to_dense().detach()
+            b_dense = b.to_dense().detach()
+            a_dense.requires_grad = True
+            b_dense.requires_grad = True
+            r1 = a_dense.matmul(b_dense)
+            r2 = torch.sparse.mm(a, b)
+
+            f1 = torch.sum(r1)
+            f1.backward()
+
+            f2 = torch.sparse.sum(r2)
+            f2.backward()
+            self.assertEqual(a.grad.to_dense(), a_dense.grad)
+            self.assertEqual(b.grad.to_dense(), b_dense.grad)
+
+        for n in range(2, 5):
```

Comment:
What about test cases for:

- uncoalesced sparse tensors
- error cases to validate the error message are thrown as expected (and don't regress)

Reply:
coalesced / uncoalesced cases are both part of the test as the test case  `TestUncoalescedSparse` extends from `TestSparse`.  `TestSparse` covers coalesce and it has  `test_sparse_matmul`.